frontend/modules/tree/treeProcessing/processing.ts

```diff
@@ -0,0 +1,419 @@
+import { intersection, keys, memoize, pickBy, uniqBy } from 'lodash';
+import { NODE_TYPES, SELECTION_STATES, VISIBILITY_STATES } from '../../../constants/tree';
+
+export class Processing {
+	public get selectedNodesIds() {
+		return keys(pickBy(this.selectionMap, (selectionState) => {
+			return selectionState !== SELECTION_STATES.UNSELECTED;
+		}));
+	}
+
+	public get fullySelectedNodesIds() {
+		return keys(pickBy(this.selectionMap, (selectionState) => {
+			return selectionState === SELECTION_STATES.SELECTED;
+		}));
+	}
+
+	public get hiddenNodesIds() {
+		return keys(pickBy(this.visibilityMap, (selectionState) => {
+			return selectionState === VISIBILITY_STATES.INVISIBLE;
+		}));
+	}
+	public nodesList = [];
+	public nodesIndexesMap = {};
+	public defaultVisibilityMap = {};
+	public meshesByModelId = {};
+	public nodesBySharedIdsMap = {};
+	public selectionMap = {};
+	public visibilityMap = {};
+
+	public getDeepChildren = memoize((node) => {
+		const nodeIndex = this.nodesIndexesMap[node._id];
+		return this.nodesList.slice(nodeIndex + 1, nodeIndex + node.deepChildrenNumber + 1);
+	}, (node) => node._id);
+
+	public getParents = memoize((node = {}) => {
+		const parents = [];
+		let nextParentId = node.parentId;
+
+		while (!!nextParentId) {
+			const parentNodeIndex = this.nodesIndexesMap[nextParentId];
+			const parentNode = this.nodesList[parentNodeIndex];
+			parents.push(parentNode);
+			nextParentId = parentNode.parentId;
+		}
+
+		return parents;
+	}, (node = {}) => node._id);
+
+	public getChildren = memoize((node = {}) => {
+		if (node.hasChildren) {
+			return this.getNodesByIds(node.childrenIds);
+		}
+
+		return [];
+	}, (node = {}) => node._id);
+
+	constructor(data) {
+		const {
+			nodesList, nodesIndexesMap, defaultVisibilityMap,
+			meshesByModelId, nodesBySharedIdsMap, visibilityMap, selectionMap
+		} = data;
+		this.nodesList = nodesList;
+		this.nodesIndexesMap = nodesIndexesMap;
+		this.defaultVisibilityMap = defaultVisibilityMap;
+		this.meshesByModelId = meshesByModelId;
+		this.visibilityMap = visibilityMap;
+		this.selectionMap = selectionMap;
+		this.nodesBySharedIdsMap = nodesBySharedIdsMap;
+	}
+
+	public clearCurrentlySelected = () => {
+		const selectedNodesIds = this.selectedNodesIds;
+		for (let index = 0, size = selectedNodesIds.length; index < size; index++) {
+			this.selectionMap[selectedNodesIds[index]] = SELECTION_STATES.UNSELECTED;
+		}
+
+		this.selectionMap = {...this.selectionMap};
+	}
+
+	public selectNodes = ({ nodesIds = [], ...extraData }) => {
+		const visibleNodesIds = nodesIds.filter((nodeId) => this.visibilityMap[nodeId] !== VISIBILITY_STATES.INVISIBLE);
```

Comment:
this line and L87  should merge together into a single loop.

there's 3  filters and getNodesByIds is another loop, resulting in O(3n) when we could do it in O(n).

Also the filter at L87 is redundant as L81 already done it?